### ISSUE TITLE
Unable to request whole CompleteState via the ControlInterface

### ISSUE THREAD

**Issue by inf17101** [eclipse-ankaios/ankaios#349]
Currently, it is not possible to fetch the whole CompleteState over the ControlInterface. On the workload site the empty CompleteState is received (workloads: None, workloadStates: None). Only parts of the CompleteState can be requested when the user sets an appropriate filter mask.

## Current Behavior

Empty CompleteState received over the ControlInterface when requesting the whole CompleteState.

## Expected Behavior

A workload shall be able to receive the whole CompleteState shall be received over the ControlInterface.

## Steps to Reproduce

1. Navigate to the control interface example folder inside the repository: cd `examples`
2. Modify the `create_request_for_complete_state` function body inside the `python_control_interface` example (adjust the filterMask to include the desiredState):
```python
    return control_api.ToAnkaios(
        request=ank_base.Request(
            completeStateRequest=ank_base.CompleteStateRequest(
                fieldMask=["desiredState", "workloadStates"] # <-- the filter mask you must change to this
            ),
            requestId=REQUEST_ID,
        )
    )
``` 
3. Change the permissions of the `startConfig.yaml` file to allow the workload to request the whole CompleteState:
```yaml
...
        - type: StateRule
          operation: Read
          filterMask:
            - "*" # change to wildcard to allow full read access
...
```
4. Run the example: `ANK_BIN_DIR=/workspaces/ankaios/target/x86_64-unknown-linux-musl/debug ./run_example.sh python_control_interface`
5. Check the log output of the workload:
```shell
podman logs -f $(podman ps -a | grep control_interface | awk '{print $1}')
```
You should see an empty CompleteState:
![image](https://github.com/user-attachments/assets/a7d20b2c-967f-4bd4-9d38-21161d0a6bed)

## Context (Environment)

Release v0.4.0
Linux
All Control Interface examples and all workloads requesting the whole CompleteState over the Control Interface are affected

## Logs

After checking the server logs (`tail -f /tmp/ankaios-server.log`) there is an issue with the tagging when using flatten and enums:
![image](https://github.com/user-attachments/assets/682c70da-eb61-4760-8855-9781abd4d6ae)

Like in the picture the error string is a serde_yaml conversion error `untagged and internally tagged enums do not support enum input`

## Final result

**To be filled by the one closing the issue.**

**Comment by inf17101** [comment]
I have analyzed the issue and I have found an issue opened on the serde_yaml part having problems since serde_yaml = "0.9" where the tagging was introduced: https://github.com/dtolnay/serde-yaml/issues/344

I know that @krucod3 was discussing that with @christoph-hamm and I know that @krucod3 left some comment in #22 about an fix to use the old still supported [singleton_map_recursive](https://docs.rs/serde_yaml/latest/serde_yaml/with/singleton_map_recursive/index.html) serializer.

I have tried to search the code base if the old singleton_map_recursive serializer fix is already included inside the Authorization changes of #301, but it is not. 

I tried to annotate each struct using enums in containers (part of the desiredState field) with the singleton_map_recursive annotation (like shown in the comment of @krucod3), but no combination has fixed the issue.

Next, I downgraded `serde_yaml` to version `0.8` and the problem is resolved. So, it is indeed the tagging introduced with serde_yaml >= 0.9.

But downgrading is not a real fix, so I decided to check if the annotations with singleton_map_recursive would fix the bug with version 0.9 when trying out the example code of the issue https://github.com/dtolnay/serde-yaml/issues/344.

And indeed when putting the `singleton_map_recursive` not on a container but on the field inside a struct using the enum which causes problems in combination with the serde flatten option, then the issue is resolved.
```rust
use serde::{Deserialize, Serialize};

fn main() {
    // this works
    let gender = Gender::Other("F22".into());
    assert_eq!(
        &gender,
        &serde_yaml::from_str(&dbg!(serde_yaml::to_string(&gender).unwrap())).unwrap()
    );

    // this still works
    let info = Info { gender };
    assert_eq!(
        &info,
        &serde_yaml::from_str(&dbg!(serde_yaml::to_string(&info).unwrap())).unwrap()
    );

    // this errors but works now because of "serde_yaml::with::singleton_map_recursive" below inside Gender struct
    let user = User { info };
    assert_eq!(
        &user,
        &serde_yaml::from_str(&dbg!(serde_yaml::to_string(&user).unwrap())).unwrap()
    );
}

#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
enum Gender {
    Female,
    Male,
    Other(String),
}

#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
struct Info {
    #[serde(with = "serde_yaml::with::singleton_map_recursive")]
    gender: Gender,
}

#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
struct User {
    #[serde(flatten)]
    info: Info,
}
```

Now, I tried to put the annotation not on containers like suggested in #22 but instead in other places on field members to get it fixed like in the example code of the other issue above. But no success, still the same issue.

Then, I decided to migrate from serde_yaml to serde_json because in the issue it was mentioned that serde_json does not have this problem and indeed, it works. There are a few modifications on the production and test code necessary, e.g. we must migrate the code of `common/src/state_manipulation/object.rs` to use serde_json instead of serde_yaml and we must change two places in the CLI code expecting now an serde_json::Error. But at the end the issue is resolved with that. The workload now gets the whole CompleteState and the tagging issue is not inside serde_json.

@krucod3: I have prepared a branch (sub-branched from release v0.4.0) with the production code changes to migrate internally to serde_json for the filtering instead of serde_yaml, since for me it looks like the serde_json is more stable and does not have the tagging problem that affects us. Please have a look if you want to migrate to the serde_json or if you want to search for an alternative solution.

**Comment by inf17101** [comment]
@krucod3: After some additional investigation I found out that the allow access rule rights are deserialized differently:

The problems are th `!stateRule` internal tags inside the sequence. The first line in the screenshot is a part of the whole current CompleteState inside `ServerState::get_complete_state_by_field_mask(...)` (line 166). It's the proto object from the current complete state after the ".into()" call. The second line is the filtered complete state inside the variable `return_state` output just before the `return_state.try_into()` call that fails when converting back the filtered Object to proto. The object.rs uses serde_yaml and when doing the filtering and inserting mappings and values it uses the tag format with (!stateRule) which confuses the serde_yaml because of the mixed tag variants used when deserializing back to proto.
![image](https://github.com/user-attachments/assets/2e984549-9bcd-4454-9b99-17db508fc01f)

**Comment by inf17101** [comment]
@krucod3: I have now also implemented a bug fix to keep the serde_yaml for parsing inside the `object.rs`. Like said in the previous comment, The conversion from proto::CompleteState to Object is different in serde_yaml = 0.9 compared to serde_yaml = 0.8. Tags are created.

The used `serde_yaml::to_value` inside the `impl TryFrom<proto::CompleteState> for Object` of `object.rs` does not use the recursive map parser. So, I have written custom code and invoked the `serde_yaml::with::singleton_map_recursive::serialize` manually.

And the problem is fixed, now I get the whole CompleteState and parts again when using the ControlInterface.

I will prepare another branch containing the fix for the serde_yaml parsing. @krucod3: Then you can compare the two variants and decide if we use serde_yaml oder serde_json in the future.

**Comment by inf17101** [comment]
After this bug fix we need a new release (v0.4.1).
Since there are already changes scheduled for v0.5.0 on the main branch, we need the concept of release branches. It could happen also in the future that we are still working on the next release and have already merged some PRs and after that we receive a bug ticket causing a new bug fix release. Because of the  issue #347 and the lifetime exceeded issues on the embedded test certificates we need a release containing the following:

- bug fix of the CompleteState filtering
- bug fix of grpc failing itests #350 

I have created the bug fix for the recursive map serializer of serde_yaml based on the v0.4.0 release tag [349_fix_complete_state](https://github.com/eclipse-ankaios/ankaios/tree/349_fix_complete_state). In addition, to unblock the CI/CD pipeline because of the failing grpc itests, I have cherry-picked the commits of #350 into the branch. The commits have only covered the itest section not production code, so I decided to go with that (and only 2 commits fixing the certificates inside the itests). With this prepared branch I have tested to create a release in my private fork and it works. The CI/CD pipeline checks out what is tagged in case of a new release build. So, we can tag release branches and build a release based on those branches as well. And with the tag it is of course traceable.

I have verified that all system tests are executed in the pipeline and I have also pulled the binaries with the tagged repo to execute the control interface examples locally to check if the whole CompleteState is still retrievable.

We still wait with the release until @krucod3 has looked over all the analysis and bug fixes. I just tested the procedure if it is possible to do releases with release branches in our current ci/cd setup.

**Comment by krucod3** [comment]
I would prefer not to have explicit mentioning of yaml in the production code.
We can fix the representation of the ControlInterfaceAccess objects during the compilation of the proto derived Rust files. 

Unfortunately the allow and deny rules were not wrapped in extra objects so I cannot properly fix the filtering of only allow and deny rules. This means that when one requests the allow rules only, also an empty deny rule would be provided. To fix this, an interface change in proto would be required, which we cannot do for a patch version update.

I opened a PR: #359

**Comment by inf17101** [comment]
> I would prefer not to have explicit mentioning of yaml in the production code. We can fix the representation of the ControlInterfaceAccess objects during the compilation of the proto derived Rust files.
> 
> Unfortunately the allow and deny rules were not wrapped in extra objects so I cannot properly fix the filtering of only allow and deny rules. This means that when one requests the allow rules only, also an empty deny rule would be provided. To fix this, an interface change in proto would be required, which we cannot do for a patch version update.
> 
> I opened a PR: #359

Sure, I agree, then it is also consistent with the other field attributes and so passed by attribute annotations.

I have already tested and reviewed the PR. Can be merged.

**Comment by krucod3** [comment]
The problem described in the issue is fixed now, but we still don't have full-fledged automated test coverage for the filtering.

On the one hand manually written utests are hard to maintain and develop, on the other fuzz-testing is not deterministic.

As the topic is out of the scope of the current issue (related to future bug prevention), I created a separate issue: #360
